cmd/initialise.go

```diff
@@ -0,0 +1,190 @@
+// THE SOFTWARE.
+
+package cmd
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/kms"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"github.com/spf13/cobra"
+)
+
+var httpClient    http.Client
+
+// InitPayload holds a Vault init request.
+type InitPayload struct {
+	RecoveryShares    int `json:"recovery_shares"`
+	RecoveryThreshold int `json:"recovery_threshold"`
+	SecretShares      int `json:"secret_shares"`
+	SecretThreshold   int `json:"secret_threshold"`
+}
+
+// InitResponse holds a Vault init response.
+type InitResponse struct {
+	Keys               []string `json:"keys"`
+	KeysBase64         []string `json:"keys_base64"`
+	RootToken          string   `json:"root_token"`
+	RecoveryKeys       []string `json:"recovery_keys"`
+	RecoveryKeysBase64 []string `json:"recovery_keys_base64"`
+}
+
+// initialiseCmd represents the initialise command
+var initialiseCmd = &cobra.Command{
+	Use:   "initialise",
+	Short: "Intitial a specified instance of Vault",
+	Long: `This command makes a requets to instance of Vault at the specified Vault address
+to initialise the instance. This command currently assumes that auto-unseal has
+been setup to occur during Vault initialisation.
+
+Once it has received the token in the response from Vault, it will encrypt and
+store this token on S3 when it can be used for authentication by entities that
+need to read from or write to the Vault instance.`,
+	Run: func(cmd *cobra.Command, args []string) {
+		// make vault init request and get root token
+		fmt.Println("Initialising Vault...")
+		rootToken := initVault()
+		fmt.Println("Initialisation complete")
+
+		// AWS setup
+		region := os.Getenv("AWS_REGION")
+		sess := session.Must(session.NewSession(&aws.Config{Region: aws.String(region)}))
+		var kmsClient = kms.New(sess, aws.NewConfig().WithRegion(region))
+		uploader := s3manager.NewUploader(sess)
+
+		// encrypt tokens with AWS KMS
+		fmt.Println("Encrypting root token...")
+		encryptedToken, errE := kmsClient.Encrypt(&kms.EncryptInput{
+			KeyId: aws.String(fullKeyID(os.Args[1], os.Args[2], region)),
+			Plaintext: []byte(rootToken),
+		})
+		checkError(errE)
+		fmt.Println("Encryption complete.")
+
+		hostname, errH := os.Hostname()
+		if errH != nil {
+			panic(errH)
+		}
+
+		tokenFileName := hostname+ "_token"
+
+		writeToFile(tokenFileName, encryptedToken.CiphertextBlob)
+
+		// upload keys to S3
+		fmt.Println("Uploading encrypted token to S3...")
+		f := openFile(tokenFileName)
+
+		s3Result, errS3 := uploader.Upload(&s3manager.UploadInput{
+			Bucket: aws.String("encrypted-tokens"),
+			Key:    aws.String(tokenFileName),
+			Body:   f,
+		})
+		checkError(errS3)
+		fmt.Println("Encrypted token successfully uploaded to S3 at", s3Result.Location)
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(initialiseCmd)
```

Comment:
Do we really need a `initialise` command?

```console
$ bin/vault-init -h
A longer description that spans multiple lines and likely contains
examples and usage of using your application. For example:

Cobra is a CLI library for Go that empowers applications.
This application is a tool to generate the needed files
to quickly create a Cobra application.

Usage:
  vault-init [command]

Available Commands:
  help        Help about any command
  initialise  Intitial a specified instance of Vault

Flags:
      --config string   config file (default is $HOME/.vault-init.yaml)
  -h, --help            help for vault-init
  -t, --toggle          Help message for toggle

Use "vault-init [command] --help" for more information about a command.
```

As is, we would need to do `vault-init initialise`, which seems unnecessary (considering the purpose of this tool is only to initialise vault), we could just do, `vault init`. Also means that a lot of what's here can go to `cmd/root.go`.

Reply:
We probably don't. we could just move it to root, particularly since we don't need to pass in arguments.

Reply:
Moved to root in https://github.com/Intellection/vault-init/pull/1/commits/f0800f6d69a05a4c616c8b4e52813195066b5328